File: Prism/Old Javascript/Events.js
//The Events Class is in charge of collecting and exporting the event data.
//It should be nothing but essentially a postal service for events to get passed
//onto Parser after receiving the call to run them.

var Events = function () {
    //var events = [];
    //
    //function cycleEventTimers(time) {
    //    for (var i = 0; i < events.length; i++) {
    //        events[i].Fire -= time;
    //    }
    //};

    //function cycleTimers() {
    //    var time = getNextTime();
    //    //Cycle up to the next event in line
    //    Timers.cycleTimers(time);
    //    cycleEventTimers(time);
    //};

    //If there are events and we're still casting, get the time difference between the current and the next one
    //function getNextTime() {
    //    if (events.length > 1 && Caster.isCasting()) {
    //        return events[1].Fire - events[0].Fire;
    //    }
    //    else {
    //        return 0;
    //    }
    //};

    //Remove the spent event
    //function finishEvent() {
    //    events.splice(0, 1);
    //};

    //Overrider
    //function removeEffectFromTarget(target, effect) {
    //    for (var i = target.Effects.length - 1; i >= 0; i--) {
    //        if (target.Effects[i][0] == effect) {
    //            target.Effects.splice(i, 1);
    //            removeTargetEventsByType(target, effect);
    //        }
    //    }
    //};
    //
    //function removeTargetEventsByType(target, effect) {
    //    for (var i = events.length - 1; i >= 0; i--) {
    //        if (events[i].Target == target.Name) {
    //            if (events[i].Name == effect) {
    //                events.splice(i, 1);
    //            }
    //        }
    //    }
    //};

    return {
        //hasEventByName: function (name) {
        //    for (var i = 0; i < events.length; i++) {
        //        if (events[i].Name === name) {
        //            return true;
        //        }
        //    }
        //    return false;
        //},
        //getEventsByName: function (name) {
        //    var eventsArray = [];
        //    for (var i = 0; i < events.length; i++) {
        //        if (events[i].Name === name) {
        //            eventsArray.push(events[i]);
        //        }
        //    }
        //    return eventsArray;
        //},
        //getEventsByTargetAndEffect: function (target, effect) {
        //    var effectArray = [];
        //    for (var i = 0; i < events.length; i++) {
        //        if (events[i].Name === effect && events[i].Target === target.Name) {
        //            effectArray.push(events[i]);
        //        }
        //    }
        //    return effectArray;
        //},
        //Used when an event is extended, such as a Pandemic effect
        //Updater: function (name, value) {
        //    for (var i = events.length - 1; i >= 0; i--) {
        //        if (events[i].Name === name) {
        //            events[i].Fire += value;
        //            events[i].Timestamp += value;
        //        }
        //    }
        //},
        //Used when a caster proc overwrites another
        //Scrubber: function (name) {
        //    for (var i = events.length - 1; i >= 0; i--) {
        //        if (events[i].Name === name) {
        //            events.splice(i, 1);
        //        }
        //    }
        //},
        //Used when an effect is consumed
        //Accelerator: function (name) {
        //    for (var i = events.length - 1; i >= 0; i--) {
        //        if (events[i].Name === name) {
        //            events[i].Timestamp = events[i].Timestamp - events[i].Fire;
        //            events[i].Fire = 0;
        //        }
        //    }
        //},
        ////Used when a target effect overwrites another
        //Overrider: function (target, spell) {
        //    removeEffectFromTarget(target, spell.Name);
        //    if (spell.Atones) {
        //        removeEffectFromTarget(target, "Atonement");
        //    }
        //},
        //Used whenever new events are added using AddEvent()
        //SortEvents: function () {
        //    events.sort(function (a, b) {
        //        if (a.Timestamp < b.Timestamp) {
        //            return -1;
        //        }
        //        if (a.Timestamp > b.Timestamp) {
        //            return 1;
        //        }
        //        return 0;
        //    });
        //},
        //Insert the new event right after the current one, to be run immediately.
        //addImmediateEvent: function (eventObj) {
        //    events.splice(1, 0, eventObj);
        //},
        //addEvent: function (eventObj) {
        //    events.push(eventObj);
        //},
        //runNextEvent: function () {
        //    Controller.switchboard(events[0]);
        //    Output.row(events[0]);
        //    cycleTimers();
        //    finishEvent();
        //}
    };
}();

File: Prism/Basic Classes/Events.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Prism.Basic_Classes
{
    public class Events
    {
        private List<CastingEvent> events;

        private void CycleTimers()
        {
            decimal time = GetNextTime();

            //Timers.CycleActiveCooldowns(time);
            CycleEventTimers(time);
        }

        private decimal GetNextTime()
        {
            if (events.Count > 1)
            {
                return events[1].Fire - events[0].Fire;
            }
            return 0;
        }

        private void CycleEventTimers(decimal time)
        {
            for (var i = 0; i < events.Count; i++)
            {
                events[i].Fire -= time;
            }
        }

        private void FinishEvent()
        {
            events.RemoveAt(0);
        }

        private void RemoveEffectFromTarget(Target target, string effect)
        {
            if (target.Effects.Exists(x => x.Name == effect))
            {
                Cooldown activeEffect = target.Effects.Find(x => x.Name == effect);
                target.Effects.Remove(activeEffect);
                RemoveTargetEventsByType(target, effect);
            }
        }

        private void RemoveTargetEventsByType(Target target, string effect)
        {
            CastingEvent thisEvent = events.Find(x => x.Target.Equals(target) && x.Name == effect);
            events.Remove(thisEvent);
        }

        public bool HasEventByName(string name)
        {
            for (int i = 0; i < events.Count; i++)
            {
                if (events[i].Name == name)
                {
                    return true;
                }
            }
            return false;
        }

        public List<CastingEvent> GetEventsByName(string name)
        {
            return events.FindAll(x => x.Name == name);
        }

        public List<CastingEvent> GetEventsByTargetAndEffect(Target target, string effect)
        {
            return events.FindAll(x => x.Target.Equals(target) && x.Name == effect);
        }

        public void PandemicExtend(string name, int value)
        {
            for (var i = events.Count - 1; i >= 0; i--)
            {
                if (events[i].Name == name)
                {
                    events[i].Fire += value;
                    events[i].Timestamp += value;
                }
            }
        }

        public void OverwriteCasterProc(string name)
        {
            for (var i = events.Count - 1; i >= 0; i--)
            {
                if (events[i].Name == name)
                {
                    events.RemoveAt(i);
                }
            }
        }

        public void AccelerateEffect(string name)
        {
            for (var i = events.Count - 1; i >= 0; i--)
            {
                if (events[i].Name == name)
                {
                    events[i].Timestamp = events[i].Timestamp - events[i].Fire;
                    events[i].Fire = 0;
                }
            }
        }

        public void OverwriteTargetEffect(Target target, Spell spell)
        {
            Cooldown activeEffect = target.Effects.Find(x => x.Name == spell.Name);
            target.Effects.Remove(activeEffect);
            if (spell.Atones)
            {
                Cooldown atonement = target.Effects.Find(x => x.Name == "Atonement");
                target.Effects.Remove(atonement);
            }
        }

        public void AddImmediateEvent(CastingEvent eventObj)
        {
            events.Insert(0, eventObj);
        }

        public void AddEvent(CastingEvent eventObj)
        {
            events.Add(eventObj);
        }

        public void SortEvents()
        {
            events.Sort((CastingEvent a, CastingEvent b) => {
                if (a.Timestamp < b.Timestamp)
                {
                    return -1;
                }
                if (a.Timestamp > b.Timestamp)
                {
                    return 1;
                }
                return 0;
            });
        }

        public void RunNextEvent()
        {
            //Controller.switchboard(events[0]);
            //Output.row(events[0]);
            CycleTimers();
            FinishEvent();
        }
    }
}
